File: my-submissions/1408-find-the-smallest-divisor-given-a-threshold/solution.cpp
class Solution {
public:
   int div_sum(vector<int>& nums,int mid){
    int sum = 0;
    for(int i = 0;i<nums.size();i++){
        sum = sum +ceil((double)(nums[i])/(double)(mid));
    }
    return sum;
}

    int smallestDivisor(vector<int>& nums, int threshold) {

        int low = 1;
        int high = *(max_element(nums.begin(),nums.end()));
        while(low<=high){
            int mid = low + (high-low)/2;
            
            if( div_sum(nums,mid)<=threshold)
            high = mid -1;
            else
            low = mid +1;
        }
        return low;
    }
 
};

File: my-submissions/1408-find-the-smallest-divisor-given-a-threshold/solution.java
class Solution {
    static int remainder(int[] a, int d){
        int res = 0;

        for(int i : a){
            res += Math.ceil((double)i/d);
        }
        return res;
    }
    public int smallestDivisor(int[] nums, int threshold) {
        int maxi = Integer.MIN_VALUE;
        for(int i : nums){
            maxi = Math.max(maxi,i);
        }

        int low = 1, high = maxi, ans = -1;

        while(low <= high){
            int mid = low + (high - low)/2;

            int res = remainder(nums,mid);

            if(res <= threshold){
                ans = mid;
                high = mid - 1;
            }
            else
            low = mid+ 1;
        }
        return ans;
    }
}
